ADBDEV-4648: Limit diskquota hash table's size according initial request

KnightMurloc opened the pull request:
Diskquota did not control the size of its hash tables in shared memory and could
have consumed shared memory not intended for it, potentially impacting the other
database subsystems. Hash tables can also grow indefinitely if the background
process on the coordinator has not started, which can happen for a number of
reasons: gone done with error, pause, isn’t started. In this case, data is not
collected from segments and some hash tables (active_tables_map, relation_cache,
relid_cache) would not be cleared and would overflow.

This patch adds a limit on the size of all hash tables in shared memory by
adding a function that checks whether the hash table is full. The function
returns HASH_FIND if the map is full and HASH_ENTER otherwise. It also report a
warning if the table is full. Implemented a GUC that controls how frequently the
warning will be reported, as it could be reported too frequently. Also added a
GUC to control size of local reject map. The size of global reject  map is set
as diskquota_max_local_reject_entries * diskquota_max_monitored_databases.

The test_active_table_limit test has been changed. Firstly, the value of
max_active_tables was changed from 2 to 5, since tables from all databases
processed by diskquota are placed in active_tables_map and with a limit of 2
tables overflow occurs even when the extension is created. Secondly, now a 
table with 10 partitions is created to overflow active_tables_map, after which
another table is created into which data is inserted that should exhaust the
quota, but since this table does not inserted into active_tables_map, its size
is not taken into account and we can insert into the table after that. At the
end, vacuum full is done to achieve the overflow of altered_reloid_cache.

red1452 committed:
Merge remote-tracking branch 'origin/2.3.0-sync2' into 2.3.0-sync

KnightMurloc committed:
Limit diskquota hash table's size according initial request

Diskquota does not control the size of its hash tables in shared memory and may
consume shared memory that is not intended for it, which may affect the other
database subsystems. Hash tables can also grow indefinitely if the background
process on the coordinator has not started, which can happen for a number of
reasons: gone done with error, pause, isn’t started.

This patch adds a limit on the size of hash tables in shared memory by adding a
wrapper over hash_search that controls the size of the table and stops adding
elements when the number of elements allocated at the beginning is reached and
report a warning. GUC is also added, which limits the frequency of report of the
warning, since in some cases it may be report too often.

RekGRpth commented on src/diskquota.c:
```diff
@@ -78,6 +78,7 @@ int  diskquota_max_workers             = 10;
 int  diskquota_max_table_segments      = 0;
 int  diskquota_max_monitored_databases = 0;
 int  diskquota_max_quota_probes        = 0;
+time_t diskquota_hashmap_overflow_report_timeout = 0;
```
why not int type?

KnightMurloc replied:
Because timestamp can be bigger than int. I will replace it with TimestampTz.

RekGRpth replied:
GUC defines int in seconds, while you may use it as 1000 * GUC milliseconds

KnightMurloc replied:
I don't think anyone will use such a large timeout. I can limit the GUC as `INT_MAX / 1000`

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,27 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+	{
+		void *result = hash_search(hashp, keyPtr, HASH_ENTER, foundPtr);
+		if (hash_get_num_entries(hashp) >= max_size &&
+		    (time(NULL) - *last_overflow_report) >= diskquota_hashmap_overflow_report_timeout)
```
Do not use time! Instead use GetCurrentTimestamp and TimestampDifferenceExceeds

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -236,7 +252,7 @@ report_altered_reloid(Oid reloid)
 
 	LWLockAcquire(diskquota_locks.altered_reloid_cache_lock, LW_EXCLUSIVE);
-	hash_search(altered_reloid_cache, &reloid, HASH_ENTER, NULL);
+	ALTERED_RELOID_CACHE_ENTER(&reloid, NULL);
```
Could you avoid macros here?

KnightMurloc replied:
I think it's better to leave it uniform with the rest of the inserts in the hash table.

RekGRpth replied:
However, I think that such macros are unintuitive because they hide the variables used internally.

RekGRpth replied:
how about such
```suggestion
	shm_hash_enter(altered_reloid_cache, &reloid, NULL, diskquota_max_active_tables, ALTERED_RELOID_CACHE_WARNING, &altered_reloid_cache_last_overflow_report, diskquota_max_active_tables);
```

KnightMurloc replied:
the task of the macro is to hide the parameters that are always the same. it may make sense to remove the macro for maps in which the insertion occurs only in one place. but it seems to me to be a complication.

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,27 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+	{
+		return hash_search(hashp, keyPtr, HASH_FIND, foundPtr);
+	}
+	else
```
After return else is redundant

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -52,6 +52,14 @@ typedef struct DiskQuotaSetOFCache
+time_t active_tables_last_overflow_report = 0;
+
+#define ACTIVE_TABLE_ENTER(keyPtr, foundPtr)                                                     \
+	shm_hash_enter(active_tables_map, keyPtr, foundPtr, diskquota_max_active_tables,             \
```
Why diskquota_max_active_tables is int, while max_size is uint?

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,27 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+// When overflowing, the warning warning_message will be report. But not more often than specified in
+// diskquota_hashmap_overflow_report_timeout. The time of the last warning is passed in last_overflow_report.
+void *
+shm_hash_enter(HTAB *hashp, void *keyPtr, bool *foundPtr, uint max_size, const char *warning_message,
```
```suggestion
shm_hash_enter(HTAB *hashp, const void *keyPtr, bool *foundPtr, uint max_size, const char *warning_message,
```

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -318,17 +334,9 @@ report_active_table_helper(const RelFileNodeBackend *relFileNode)
 
 	LWLockAcquire(diskquota_locks.active_table_lock, LW_EXCLUSIVE);
-	entry = hash_search(active_tables_map, &item, HASH_ENTER_NULL, &found);
+	entry = ACTIVE_TABLE_ENTER(&item, &found);
```
Why did you replace HASH_ENTER_NULL by HASH_ENTER?

KnightMurloc replied:
shm_hash_enter works as HASH_ENTER_NULL. with the exception that it returns NULL not when the memory has run out, but when the limit is reached and entry not found.

RekGRpth replied:
> it returns NULL not when the memory has run out

What is happen if memory has run out?

KnightMurloc replied:
we allocate the maximum number of elements when creating a table, so there should be no memory allocations during execution.

https://github.com/arenadata/gpdb/blob/215d8d8f9fec39300e4eb84977ade4fce040ebc1/src/backend/utils/hash/dynahash.c#L442-L457

RekGRpth replied:
> we allocate the maximum number of elements when creating a table

Why is not used flag HASH_FIXED_SIZE in this case?

RekGRpth replied:
> so there should be no memory allocations during execution

I do not think so: ShmemInitHash has init_size and max_size arguments, and init_size is passed to hash_create as nelem - maximum number of elements expected, not max_size!

KnightMurloc replied:
in our case, `init_size == max_size`. and we should not go beyond this limit.
https://github.com/arenadata/diskquota/blob/e6899aaac6dc36d095c740d8e09c99b326a189fa/src/gp_activetable.c#L106-L107

RekGRpth replied:
I think in this case we may use HASH_FIXED_SIZE flag

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,27 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+shm_hash_enter(HTAB *hashp, void *keyPtr, bool *foundPtr, uint max_size, const char *warning_message,
+               time_t *last_overflow_report)
+{
+	if (hash_get_num_entries(hashp) >= max_size)
```
I think you should some lock due to shared hash

KnightMurloc replied:
the hash_search call was already done with locks.

KnightMurloc committed:
using functions to work with time from gpdb instead of libc. add flag HASH_FIXED_SIZE to all hash tables in shared memory.

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1678,32 +1678,26 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
-	if (hash_get_num_entries(hashp) >= max_size)
+	void *result = hash_search(hashp, keyPtr, HASH_ENTER_NULL, foundPtr);
+	if (hash_get_num_entries(hashp) >= max_size &&
+	    TimestampDifferenceExceeds(*last_overflow_report, GetCurrentTimestamp(),
```
try to avoid call GetCurrentTimestamp twice: save it to local var and then use it

RekGRpth commented:
do we steel need a variable diskquota_table_size_entry_num?

RekGRpth commented on src/quotamodel.c:
```diff
@@ -582,8 +582,8 @@ init_disk_quota_model(uint32 id)
 	hash_ctl.entrysize = sizeof(TableSizeEntry);
-	table_size_map     = DiskquotaShmemInitHash(str.data, INIT_NUM_TABLE_SIZE_ENTRIES, MAX_NUM_TABLE_SIZE_ENTRIES,
-	                                            &hash_ctl, HASH_ELEM, DISKQUOTA_TAG_HASH);
+	table_size_map = DiskquotaShmemInitHash(str.data, MAX_NUM_TABLE_SIZE_ENTRIES, MAX_NUM_TABLE_SIZE_ENTRIES, &hash_ctl,
```
now, for each database we allocate map with size 10 * 1024 * 1024 / 100 by default?

KnightMurloc replied:
Yes. How much do we want to reduce memory consumption at the start? Because otherwise we won't be able to use HASH_FIXED_SIZE.

RekGRpth replied:
So, after all, it is possible to initialize a small shared map at startup, and then increase its size?

KnightMurloc replied:
yes, but then we will need to control the size as it was in the first version of the patch.

RekGRpth replied:
I think it would better

RekGRpth commented:
why not using new approach [here](https://github.com/arenadata/diskquota/pull/28/commits/705d2ca0bd614a11201a2cb384caf832f718cb39#diff-74478810938875c96024185d7ae83994b0aa5aaf71f9398ef87a03d5dfd2d9d7R1044)?

KnightMurloc commented:
I used it in the first version of the patch. I just moved it into a separate function to reduce code duplication.

KnightMurloc committed:
Revert "using functions to work with time from gpdb instead of libc. add flag HASH_FIXED_SIZE to all hash tables in shared memory."

This reverts commit 705d2ca0bd614a11201a2cb384caf832f718cb39.

KnightMurloc committed:
using functions to work with time from gpdb instead of libc

RekGRpth commented:
> do we steel need a variable diskquota_table_size_entry_num?

and diskquota_quota_info_entry_num

KnightMurloc commented:
we don't need them. I will remake all the tables in shared memory in the same style.

KnightMurloc committed:
unified logic for adding elements to hash tables in shared memory for all tables

RekGRpth commented on src/diskquota.c:
```diff
@@ -1802,15 +1801,6 @@ init_launcher_shmem()
 		}
 	}
-	/* init TableSizeEntry counter */
-	diskquota_table_size_entry_num =
```
Could we decrease shmem on sizeof(pg_atomic_uint32) twice?

KnightMurloc replied:
the counter `diskquota_quota_info_entry_num` is not taken into account when allocating shared memory. I have already deleted the counter for table_size.

RekGRpth replied:
> the counter `diskquota_quota_info_entry_num` is not taken into account when allocating shared memory. I have already deleted the counter for table_size.

https://github.com/arenadata/diskquota/pull/28/files#diff-4185ed65003844abdc2e3793b5804476cb289c83c14df030a4474dcc59c5192cL1811-L1812

RekGRpth replied:
> I have already deleted the counter for table_size.

OK

KnightMurloc replied:
I mean its size is not added to the size of the shared memory https://github.com/greenplum-db/diskquota/commit/547e5d18fa72cd7d26fe11b4073e07b493a65471#diff-74478810938875c96024185d7ae83994b0aa5aaf71f9398ef87a03d5dfd2d9d7R591

RekGRpth replied:
Forgotten?

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,27 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+	{
+		return hash_search(hashp, keyPtr, HASH_FIND, foundPtr);
+	}
+	void *result = hash_search(hashp, keyPtr, HASH_ENTER, foundPtr);
```
Could you put declaration to begin of block?

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,27 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+	void *result = hash_search(hashp, keyPtr, HASH_ENTER, foundPtr);
+
+	TimestampTz current_time = GetCurrentTimestamp();
+	if (hash_get_num_entries(hashp) >= max_size &&
```
I suggest to do one call GetCurrentTimestamp only if hash_get_num_entries(hashp) >= max_size

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -52,6 +52,14 @@ typedef struct DiskQuotaSetOFCache
 } DiskQuotaSetOFCache;
 
 HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
+time_t active_tables_last_overflow_report = 0;
```
```suggestion
TimestampTz active_tables_last_overflow_report = 0;
```

RekGRpth replied:
not only here!

RekGRpth commented on src/relation_cache.c:
```diff
@@ -177,10 +193,19 @@ update_relation_cache(Oid relid)
-
-	relid_entry = hash_search(relid_cache, &relid_entry_data.relfilenode, HASH_ENTER, NULL);
+	relid_entry = RELID_CACHE_ENTER(&relid_entry_data.relfilenode, NULL);
+	if (relid_entry == NULL)
```
```suggestion
	if (relid_entry != NULL) *relid_entry = relid_entry_data;
```

KnightMurloc replied:
Are you sure we can do this? Because if there is no space in relation_cache, then there is probably no space in active_tables_map and diskquota won't work correctly anymore.

RekGRpth replied:
why not? it is optimized (by compiler) version of memcpy only

RekGRpth replied:
we may also do so
```suggestion
	if (relid_entry != NULL) memcpy(relid_entry, &relid_entry_data, sizeof(DiskQuotaRelidCacheEntry));
```

KnightMurloc replied:
I mean, this is not just a check before copying if there is no space, we exit the function. how much do we want to continue executing the function if there is no place for the hash map? and how correct would that be?

RekGRpth replied:
You may add something like
```c
	if (relation_entry == NULL) return;
	prelid = get_primary_table_oid(relid, false);
```

KnightMurloc replied:
I don't see the point in searching in the second map if we didn't find it in the first one. since we add entries in pairs, if we didn't find them in the first one, we probably won't find them in the second one either.

RekGRpth replied:
May be put them into one map?

KnightMurloc replied:
But we look for them separately. also, I don't want to do more refactoring unrelated to the original problem.

RekGRpth reviewed (commented):
You can remove extra parentheses if there is only one instruction in the block.

KnightMurloc committed:
few fixes

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1691,19 +1691,22 @@ void *
 	{
-		ereport(WARNING, (errmsg(warning_message, guc_value)));
-		*last_overflow_report = current_time;
+		current_time = GetCurrentTimestamp();
```
```suggestion
		TimestampTz current_time = GetCurrentTimestamp();
```

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1691,19 +1691,22 @@ void *
+	void       *result;
+	TimestampTz current_time;
+
+	if (hash_get_num_entries(hashp) >= max_size) return hash_search(hashp, keyPtr, HASH_FIND, foundPtr);
```
```suggestion
	if (hash_get_num_entries(hashp) < max_size) return hash_search(hashp, keyPtr, HASH_ENTER, foundPtr);
	if (hash_get_num_entries(hashp) ==  max_size)
	{
		TimestampTz current_time = GetCurrentTimestamp();
		if (TimestampDifferenceExceeds(*last_overflow_report, current_time,
		                               diskquota_hashmap_overflow_report_timeout * 1000))
		{
			ereport(WARNING, (errmsg(warning_message, guc_value)));
			*last_overflow_report = current_time;
		}
	}
	return hash_search(hashp, keyPtr, HASH_FIND, foundPtr);
```

KnightMurloc committed:
remake shm_hash_enter

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -61,6 +69,14 @@ HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
  * dbid will be removed from it when droping diskquota extension
  */
 HTAB *altered_reloid_cache = NULL; // Set<Oid>
+time_t altered_reloid_cache_last_overflow_report = 0;
```
```suggestion
TimestampTz altered_reloid_cache_last_overflow_report = 0;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -185,11 +191,34 @@ struct LocalRejectMapEntry
 static HTAB *disk_quota_reject_map       = NULL;
 static HTAB *local_disk_quota_reject_map = NULL;
 
+static time_t disk_quota_reject_last_overflow_report       = 0;
```
```suggestion
static TimestampTz disk_quota_reject_last_overflow_report = 0;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -185,11 +191,34 @@ struct LocalRejectMapEntry
 static HTAB *local_disk_quota_reject_map = NULL;
 
+static time_t disk_quota_reject_last_overflow_report       = 0;
+static time_t local_disk_quota_reject_last_overflow_report = 0;
```
```suggestion
static TimestampTz local_disk_quota_reject_last_overflow_report = 0;
```

RekGRpth commented on src/relation_cache.c:
```diff
@@ -32,6 +32,22 @@
 HTAB *relation_cache = NULL;
 HTAB *relid_cache    = NULL;
 
+extern time_t active_tables_last_overflow_report;
```
```suggestion
extern TimestampTz active_tables_last_overflow_report;
```

KnightMurloc committed:
use TimestampTz instead of time_t

RekGRpth commented:
update branch and resolve possible conflicts

KnightMurloc committed:
Merge branch 'gpdb' into ADBDEV-4648

KnightMurloc committed:
fix formating

KnightMurloc committed:
fix formating

KnightMurloc committed:
make tests more stable

red1452 commented on src/diskquota.c:
```diff
@@ -70,14 +70,15 @@ static volatile sig_atomic_t got_sigusr1 = false;
+int  diskquota_max_table_segments              = 0;
+int  diskquota_max_monitored_databases         = 0;
+int  diskquota_max_quota_probes                = 0;
+int  diskquota_hashmap_overflow_report_timeout = 0;
```
Are you sure that this refactoring is needed here? I think this patch must contain only fixes. It will help us to minimize conflicts in the future.

KnightMurloc replied:
clangd-format requires all variables to be aligned. since a long variable was added, the rest had to be aligned.

red1452 commented on src/diskquota.c:
```diff
@@ -414,6 +409,10 @@ define_guc_variables(void)
 	                        &diskquota_max_quota_probes, 1024 * 1024, 1024 * INIT_QUOTA_MAP_ENTRIES, INT_MAX,
 	                        PGC_POSTMASTER, 0, NULL, NULL, NULL);
+	DefineCustomIntVariable("diskquota.hashmap_overflow_report_timeout",
+	                        "Time interval in seconds between shared hash map overflow report.", NULL,
```
It is better to use the same style for help message as `naptime` or `worker_timeout`.

red1452 commented on src/quotamodel.c:
```diff
@@ -184,12 +190,35 @@ struct LocalRejectMapEntry
+#define REJECT_MAP_ENTER(keyPtr, foundPtr)                                                 \
+	shm_hash_enter(disk_quota_reject_map, keyPtr, foundPtr, MAX_DISK_QUOTA_REJECT_ENTRIES, \
+	               "[diskquota] Shared disk quota reject map size limit reached (%d)."     \
+	               "Some out-of-limit schemas or roles will be lost"                       \
+	               "in rejectmap.",                                                        \
```
```suggestion
	               "Some out-of-limit schemas or roles will be lost in rejectmap."         \
```

red1452 commented on src/quotamodel.c:
```diff
@@ -184,12 +190,35 @@ struct LocalRejectMapEntry
+
+#define REJECT_MAP_ENTER(keyPtr, foundPtr)                                                 \
+	shm_hash_enter(disk_quota_reject_map, keyPtr, foundPtr, MAX_DISK_QUOTA_REJECT_ENTRIES, \
+	               "[diskquota] Shared disk quota reject map size limit reached (%d)."     \
```
```suggestion
	               "[diskquota] Shared disk quota reject map size limit reached (%d). "    \
```

KnightMurloc committed:
change guc description. fix warning message

KnightMurloc committed:
rework

KnightMurloc committed:
fix format

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,25 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+ * Returns HASH_FIND if hash table is full and HASH_ENTER otherwise.
+ */
+HASHACTION
+check_hash_fullness(HTAB *hashp, int max_size, const char *warning_message, TimestampTz *last_overflow_report,
```
Can this function have a prototype like this?
```suggestion
check_hash_fullness(HTAB *hashp, int max_size, TimestampTz *last_overflow_report, const char *warning_message, ...)
```

KnightMurloc replied:
we cannot pass variadic arguments to another function. But it may be worth formatting the string once during initialization, since changing the GUC requires a restart. what do you think?

RekGRpth replied:
> changing the GUC requires a restart

all used GUCs?

RekGRpth replied:
I came up with a better solution! See below.

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -236,7 +248,10 @@ report_altered_reloid(Oid reloid)
 
 	LWLockAcquire(diskquota_locks.altered_reloid_cache_lock, LW_EXCLUSIVE);
-	hash_search(altered_reloid_cache, &reloid, HASH_ENTER, NULL);
+	HASHACTION action =
```
Is it possible to mix declaration and code in a diskquota?

KnightMurloc replied:
this is already used in a lot of places in diskquota.

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -857,7 +866,9 @@ get_active_tables_oid(void)
 	while ((active_table_file_entry = (DiskQuotaActiveTableFileEntry *)hash_seq_search(&iter)) != NULL)
 	{
 		/* TODO: handle possible ERROR here so that the bgworker will not go down. */
```
Has this already been done there?

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,25 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+		if (TimestampDifferenceExceeds(*last_overflow_report, current_time,
+		                               diskquota_hashmap_overflow_report_timeout * 1000))
+		{
+			ereport(WARNING, (errmsg(warning_message, guc_value)));
```
remove `%d` from warning_message and
```suggestion
			ereport(WARNING, (errmsg("%s %d", warning_message, guc_value)));
```

KnightMurloc replied:
But what does it change?

RekGRpth replied:
with this change warning_message is not string format template

KnightMurloc replied:
I don't see any improvements except that now the GUC value will always be at the end.

RekGRpth replied:
The current version of the check_hash_fullness function can be passed a string with multiple substitutions to format the message, which will result in a runtime error.

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,25 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+check_hash_fullness(HTAB *hashp, int max_size, const char *warning_message, TimestampTz *last_overflow_report,
+                    int guc_value)
+{
+	if (hash_get_num_entries(hashp) < max_size) return HASH_ENTER;
```
```suggestion
	int num_entries = hash_get_num_entries(hashp);

	if (num_entries < max_size) return HASH_ENTER;

	if (num_entries == max_size)
```

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,25 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+
+/*
+ * Returns HASH_FIND if hash table is full and HASH_ENTER otherwise.
+ */
```
```suggestion
 * It can be used only under lock.
 */
```

KnightMurloc committed:
change warning messages

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,28 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+
+	if (num_entries == max_size)
+	{
+		TimestampTz current_time = GetCurrentTimestamp();
```
add empty line for beautify
```suggestion
		TimestampTz current_time = GetCurrentTimestamp();

```

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,28 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+			ereport(WARNING, (errmsg("%s %d", warning_message, guc_value)));
+			*last_overflow_report = current_time;
+		}
+	}
```
add empty line for beautify
```suggestion
	}

```

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -51,7 +51,13 @@ typedef struct DiskQuotaSetOFCache
+HTAB       *active_tables_map                  = NULL; // Set<DiskQuotaActiveTableFileEntry>
+TimestampTz active_tables_last_overflow_report = 0;
+
+#define ACTIVE_TABLE_WARNING                                                      \
```
```suggestion
#define ACTIVE_TABLES_MAP_WARNING                                                      \
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -144,7 +142,13 @@ typedef enum
+HTAB       *quota_info_map;
+TimestampTz quota_info_map_last_overflow_report = 0;
+
+#define QUOTA_INFO_WARNING                                             \
```
```suggestion
#define QUOTA_INFO_MAP_WARNING                                             \
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
+static HTAB       *table_size_map                  = NULL;
+static TimestampTz table_size_last_overflow_report = 0;
+
+#define TABLE_SIZE_WARNING                                                 \
```
```suggestion
#define TABLE_SIZE_MAP_WARNING                                                 \
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
+static TimestampTz disk_quota_reject_last_overflow_report       = 0;
+static TimestampTz local_disk_quota_reject_last_overflow_report = 0;
+
+#define REJECT_MAP_WARNING                                          \
```
```suggestion
#define DISK_QUOTA_REJECT_MAP_WARNING                                          \
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
+	"[diskquota] Shared disk quota reject map size limit reached. " \
+	"Some out-of-limit schemas or roles will be lost in rejectmap. Current limit:"
+
+#define LOCAL_REJECT_MAP_WARNING "[diskquota] the number of local reject map entries reached the limit:"
```
```suggestion
#define LOCAL_DISK_QUOTA_REJECT_MAP_WARNING "[diskquota] the number of local reject map entries reached the limit:"
```

KnightMurloc committed:
change macro names

red1452 commented on tests/regress/expected/test_activetable_limit.out:
```diff
@@ -26,31 +27,47 @@ SELECT diskquota.wait_for_worker_new_epoch();
+WARNING:  [diskquota] the number of active tables reached the limit, please increase the GUC value for diskquota.max_active_tables. Current diskquota.max_active_tables value: 5
+WARNING:  [diskquota] the number of active tables reached the limit, please increase the GUC value for diskquota.max_active_tables. Current diskquota.max_active_tables value: 5  (seg0 127.0.1.1:6002 pid=281310)
+WARNING:  [diskquota] the number of active tables reached the limit, please increase the GUC value for diskquota.max_active_tables. Current diskquota.max_active_tables value: 5  (seg1 127.0.1.1:6003 pid=281312)
+WARNING:  [diskquota] the number of active tables reached the limit, please increase the GUC value for diskquota.max_active_tables. Current diskquota.max_active_tables value: 5  (seg2 127.0.1.1:6004 pid=281311)
```
WARNING messages from segments are ignored at test (I don't know why). I think it is important for this test.
```diff
--- \/home\/gpadmin\/diskquota_artifacts\/tests\/regress\/expected\/test_activetable_limit\.out	2023-12-20 18:11:48.210422362 +0000
+++ \/home\/gpadmin\/diskquota_artifacts\/tests\/regress\/results\/test_activetable_limit\.out	2023-12-20 18:11:48.214422455 +0000
@@ -31,7 +31,6 @@
 -- We create twice as many tables as the limit to ensure that the active_tables table is overflow.
 CREATE TABLE s.t1 (a int, b int) DISTRIBUTED BY (a)
     PARTITION BY RANGE (b) ( START (0) END (10) EVERY (1) );
-WARNING:  [diskquota] the number of active tables reached the limit, please increase the GUC value for diskquota.max_active_tables. Current diskquota.max_active_tables value: 5
 CREATE TABLE s.t2(i int) DISTRIBUTED BY (i);
 INSERT INTO s.t2 SELECT generate_series(1, 100000);
 SELECT diskquota.wait_for_worker_new_epoch();
@@ -41,15 +40,15 @@
 (1 row)
 
 INSERT INTO s.t1 SELECT a, a from generate_series(0, 9)a; -- should be successful
+ERROR:  schema's disk space quota exceeded with name: s
 select count(*) from s.t1;
  count 
 -------
-    10
+     0
 (1 row)
 
 -- altered reloid cache overflow check. expected warning.
 vacuum full;
-WARNING:  [diskquota] the number of altered reloid cache entries reached the limit, please increase the GUC value for diskquota.max_active_tables. Current diskquota.max_active_tables value: 5
 DROP EXTENSION diskquota;
 \c contrib_regression
 DROP DATABASE test_tablenum_limit_01;

======================================================================
```

KnightMurloc committed:
use an external table to check warnings in the test.

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -51,7 +51,13 @@ typedef struct DiskQuotaSetOFCache
 
-HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
+HTAB       *active_tables_map                  = NULL; // Set<DiskQuotaActiveTableFileEntry>
+TimestampTz active_tables_last_overflow_report = 0;
```
```suggestion
TimestampTz active_tables_map_last_overflow_report = 0;
```

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,30 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+	{
+		TimestampTz current_time = GetCurrentTimestamp();
+
+		if (TimestampDifferenceExceeds(*last_overflow_report, current_time,
```
```suggestion
		if (*last_overflow_report == 0 || TimestampDifferenceExceeds(*last_overflow_report, current_time,
```

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -51,7 +51,13 @@ typedef struct DiskQuotaSetOFCache
+#define ACTIVE_TABLES_MAP_WARNING                                                 \
+	"[diskquota] the number of active tables reached the limit, please increase " \
+	"the GUC value for diskquota.max_active_tables. Current "                     \
+	"diskquota.max_active_tables value:"
```
I think that the current value in the warning can be omitted; this will greatly simplify the functions.

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -60,7 +66,13 @@ HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
  */
-HTAB *altered_reloid_cache = NULL; // Set<Oid>
+HTAB       *altered_reloid_cache                      = NULL; // Set<Oid>
+TimestampTz altered_reloid_cache_last_overflow_report = 0;
```
```suggestion
static TimestampTz altered_reloid_cache_last_overflow_report = 0;
```

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -60,7 +66,13 @@ HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
  * dbid will be removed from it when droping diskquota extension
  */
-HTAB *altered_reloid_cache = NULL; // Set<Oid>
+HTAB       *altered_reloid_cache                      = NULL; // Set<Oid>
```
```suggestion
static HTAB       *altered_reloid_cache                      = NULL; // Set<Oid>
```

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -51,7 +51,13 @@ typedef struct DiskQuotaSetOFCache
 } DiskQuotaSetOFCache;
 
-HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
+HTAB       *active_tables_map                  = NULL; // Set<DiskQuotaActiveTableFileEntry>
```
```suggestion
static HTAB       *active_tables_map                  = NULL; // Set<DiskQuotaActiveTableFileEntry>
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -144,7 +142,13 @@ typedef enum
 Oid    quota_key_caches[NUM_QUOTA_TYPES][MAX_NUM_KEYS_QUOTA_MAP] = {
         {NAMESPACEOID}, {AUTHOID}, {NAMESPACEOID, TABLESPACEOID}, {AUTHOID, TABLESPACEOID}, {TABLESPACEOID}};
-HTAB *quota_info_map;
+HTAB       *quota_info_map;
```
```suggestion
static HTAB       *quota_info_map;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -144,7 +142,13 @@ typedef enum
         {NAMESPACEOID}, {AUTHOID}, {NAMESPACEOID, TABLESPACEOID}, {AUTHOID, TABLESPACEOID}, {TABLESPACEOID}};
-HTAB *quota_info_map;
+HTAB       *quota_info_map;
+TimestampTz quota_info_map_last_overflow_report = 0;
```
```suggestion
static TimestampTz quota_info_map_last_overflow_report = 0;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
 /* using hash table to support incremental update the table size entry.*/
-static HTAB *table_size_map = NULL;
+static HTAB       *table_size_map                  = NULL;
+static TimestampTz table_size_last_overflow_report = 0;
```
```suggestion
static TimestampTz table_size_map_last_overflow_report = 0;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
 static HTAB *disk_quota_reject_map       = NULL;
 static HTAB *local_disk_quota_reject_map = NULL;
 
+static TimestampTz disk_quota_reject_last_overflow_report       = 0;
```
```suggestion
static TimestampTz disk_quota_reject_map_last_overflow_report       = 0;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
 static HTAB *local_disk_quota_reject_map = NULL;
 
+static TimestampTz disk_quota_reject_last_overflow_report       = 0;
+static TimestampTz local_disk_quota_reject_last_overflow_report = 0;
```
```suggestion
static TimestampTz local_disk_quota_reject_map_last_overflow_report = 0;
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +188,27 @@ struct LocalRejectMapEntry
+
+#define DISK_QUOTA_REJECT_MAP_WARNING                               \
+	"[diskquota] Shared disk quota reject map size limit reached. " \
+	"Some out-of-limit schemas or roles will be lost in rejectmap. Current limit:"
```
May be add info for increase some GUC?

KnightMurloc replied:
the disk_quota_reject_map limit is fixed MAX_DISK_QUOTA_REJECT_ENTRIES

RekGRpth commented on tests/regress/sql/test_activetable_limit.sql:
```diff
@@ -19,29 +20,34 @@ INSERT INTO a03 values(generate_series(0, 500));
-INSERT INTO s.t1 SELECT generate_series(1, 100000); -- ok. diskquota soft limit does not check when first write
+-- We create twice as many tables as the limit to ensure that the active_tables table is overflow.
+CREATE TABLE s.t1 (a int, b int) DISTRIBUTED BY (a)
+    PARTITION BY RANGE (b) ( START (0) END (10) EVERY (1) );
```
May we suppress notices?

KnightMurloc replied:
these notices should already be suppressed.
https://github.com/arenadata/diskquota/blob/5e402a7475cc6bfaee295ea0f49a0c3e753d2d06/tests/init_file#L10

RekGRpth replied:
see tests/regress/expected/test_activetable_limit.out

KnightMurloc replied:
The test_many_active_table and test_partion tests also have these notices.

RekGRpth replied:
you may increase log level to warning before create table with partitions and reset after

KnightMurloc replied:
I don't see the point in doing this since all the tests leave these notices as is.

KnightMurloc committed:
rename waribles. remove GUC value from warning messages.

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1700,10 +1699,10 @@ check_hash_fullness(HTAB *hashp, int max_size, const char *warning_message, Time
+		                                                             diskquota_hashmap_overflow_report_timeout * 1000))
 		{
-			ereport(WARNING, (errmsg("%s %d", warning_message, guc_value)));
+			ereport(WARNING, (errmsg(warning_message)));
```
```suggestion
			ereport(WARNING, (errmsg("%s", warning_message)));
```
to avoid substitution
or even so
```suggestion
			ereport(WARNING, (errmsg("[diskquota] %s", warning_message)));
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -188,26 +187,25 @@ struct LocalRejectMapEntry
 #define DISK_QUOTA_REJECT_MAP_WARNING                               \
 	"[diskquota] Shared disk quota reject map size limit reached. " \
-	"Some out-of-limit schemas or roles will be lost in rejectmap. Current limit:"
+	"Some out-of-limit schemas or roles will be lost in rejectmap."
```
Maybe add a GUC(s) for these limits, otherwise why issue a warning if nothing can be done about it at all?

KnightMurloc committed:
Add guc diskquota.max_reject_entries that control size of
local_disk_quota_reject_map. The size of disk_quota_reject_map is calculated as
the maximum size of the local table multiplied by the maximum number of
databases. The variables that store the time of the last report on reaching the
limit are now stored in shared memory to prevent spam from dynamic workers.

KnightMurloc committed:
fix format

RekGRpth commented on src/quotamodel.c:
```diff
@@ -184,12 +183,28 @@ struct LocalRejectMapEntry
 /* using hash table to support incremental update the table size entry.*/
-static HTAB *table_size_map = NULL;
+static HTAB        *table_size_map                      = NULL;
+static TimestampTz *table_size_map_last_overflow_report = 0;
```
Why did you change the type of the variable to a pointer?

KnightMurloc replied:
because it is now stored in shared memory. I forgot to change 0 to NULL.

RekGRpth replied:
I mean why

> it is now stored in shared memory

?

KnightMurloc replied:
because otherwise its value is not saved for dynamic workers, which leads to spam warnings. especially during tests.

RekGRpth replied:
may be add some comments to code, why in one place tz is pointer while in another - not?

KnightMurloc committed:
change default value of table_size_map_last_overflow_report from 0 to NULL

KnightMurloc committed:
add comment.

KnightMurloc committed:
fix format

RekGRpth commented on src/quotamodel.c:
```diff
@@ -540,10 +532,11 @@ DiskQuotaShmemSize(void)
 		size = add_size(size, diskquota_worker_shmem_size() * diskquota_max_monitored_databases);
 		size = add_size(size, hash_estimate_size(MAX_QUOTA_MAP_ENTRIES, sizeof(QuotaInfoEntry)) *
 		                              diskquota_max_monitored_databases);
+		size = add_size(size,
```
Why is this line not in the diskquota_worker_shmem_size function?

KnightMurloc replied:
because, the size of the quota_info_map hash table is added here.

RekGRpth replied:
It would be more logical to move this size to that function.

KnightMurloc committed:
fix test

KnightMurloc committed:
move quota_info_map_last_overflow_report size adding to diskquota_worker_shmem_size

RekGRpth commented on src/diskquota_utility.c:
```diff
@@ -1683,3 +1683,29 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
+		if (*last_overflow_report == 0 || TimestampDifferenceExceeds(*last_overflow_report, current_time,
+		                                                             diskquota_hashmap_overflow_report_timeout * 1000))
+		{
+			ereport(WARNING, (errmsg("%s", warning_message)));
```
```suggestion
			ereport(WARNING, (errmsg("[diskquota] %s", warning_message)));
```

RekGRpth commented on src/quotamodel.c:
```diff
@@ -540,7 +533,6 @@ DiskQuotaShmemSize(void)
 		size = add_size(size, diskquota_launcher_shmem_size());
-		size = add_size(size, sizeof(pg_atomic_uint32));
 		size = add_size(size, diskquota_worker_shmem_size() * diskquota_max_monitored_databases);
 		size = add_size(size, hash_estimate_size(MAX_QUOTA_MAP_ENTRIES, sizeof(QuotaInfoEntry)) *
```
May be, move this line to the diskquota_worker_shmem_size function too?

RekGRpth commented on src/relation_cache.c:
```diff
@@ -32,6 +32,16 @@
+	"the GUC value for diskquota.max_active_tables."
+
+#define RELID_CACHE_WARNING                                                                \
+	"[diskquota] the number of relation cache entries reached the limit, please increase " \
```
```suggestion
	"[diskquota] the number of relid cache entries reached the limit, please increase " \
```

KnightMurloc committed:
change warning messages. move size adding of quota_info_map to diskquota_worker_shmem_size

RekGRpth commented:
check tests

KnightMurloc committed:
make tests more stable

RekGRpth commented on tests/regress/sql/test_activetable_limit.sql:
```diff
@@ -19,29 +20,34 @@ INSERT INTO a03 values(generate_series(0, 500));
 \c test_tablenum_limit_02
 CREATE EXTENSION diskquota;
+
+CREATE EXTERNAL WEB TABLE segment_logs(line text)
```
Why not use gp_toolkit.gp_log_system?

RekGRpth commented on tests/regress/sql/test_rejectmap_limit.sql:
```diff
@@ -0,0 +1,54 @@
+DROP DATABASE test_reject_map_limit_01;
+
+\! gpconfig -r diskquota.max_reject_entries > /dev/null
+\! gpstop -arf > /dev/null
```
```suggestion
\! gpstop -arf > /dev/null

```

KnightMurloc committed:
add comment to tests. add new line at the end of new test

Commit:
Merge branch 'gpdb' into ADBDEV-4648

KnightMurloc committed:
change macro to const varibles.

KnightMurloc committed:
make warning constats static

RekGRpth commented on src/gp_activetable.c:
```diff
@@ -60,7 +65,12 @@ HTAB *active_tables_map = NULL; // Set<DiskQuotaActiveTableFileEntry>
+static HTAB       *altered_reloid_cache                      = NULL; // Set<Oid>
+static TimestampTz altered_reloid_cache_last_overflow_report = 0;
+
+const char *altered_reloid_cache_warning =
```
```suggestion
static const char *altered_reloid_cache_warning =
```

KnightMurloc committed:
add static to altered_reloid_cache_warning
